File: mcp-server/python/algovoi_mcp/redact.py
"""
Output sanitisation for tool responses.

Two things happen before any response returns to the LLM:

  1. Sensitive keys are redacted — ``mnemonic``, ``private_key``, ``secret``,
     ``api_key``, ``webhook_secret``, ``authorization``, ``access_token``,
     ``refresh_token``, ``bearer_token``, ``password``.
     (The *checkout* ``token`` returned by ``create_payment_link`` is NOT in
     the list — it's a short opaque ID the caller needs, not a credential.)

  2. String fields are truncated to :data:`MAX_STR` to defend against
     prompt-injection payloads embedded in attacker-controlled blockchain
     data (transaction memos, NFT metadata, labels).  See §4.4 of
     ALGOVOI_MCP.md.
"""

from __future__ import annotations

from typing import Any

MAX_STR = 512

SENSITIVE_KEYS: frozenset[str] = frozenset(
    {
        "mnemonic",
        "private_key",
        "privatekey",
        "secret",
        "api_key",
        "apikey",
        "password",
        "passwd",
        "authorization",
        "auth",
        "webhook_secret",
        "webhooksecret",
        "access_token",
        "refresh_token",
        "bearer_token",
        # Env-var names too, in case something forwards them
        "algovoi_api_key",
        "algovoi_webhook_secret",
    }
)

_REDACTED = "[REDACTED]"


def scrub(obj: Any) -> Any:
    """
    Recursively sanitise ``obj`` for safe inclusion in tool responses.

    - Dict keys matched (case-insensitive, exact match) against
      :data:`SENSITIVE_KEYS` have their values replaced with ``"[REDACTED]"``.
    - String values longer than :data:`MAX_STR` are truncated with a visible
      ``"... [truncated N chars]"`` suffix.
    - Lists and nested dicts are walked recursively.
    - All other types pass through unchanged.
    """
    if isinstance(obj, dict):
        out: dict[Any, Any] = {}
        for k, v in obj.items():
            if isinstance(k, str) and k.lower() in SENSITIVE_KEYS:
                out[k] = _REDACTED
            else:
                out[k] = scrub(v)
        return out
    if isinstance(obj, list):
        return [scrub(x) for x in obj]
    if isinstance(obj, tuple):
        return tuple(scrub(x) for x in obj)
    if isinstance(obj, str) and len(obj) > MAX_STR:
        return obj[:MAX_STR] + f"... [truncated {len(obj) - MAX_STR} chars]"
    return obj

File: mcp-server/typescript/src/redact.ts
/**
 * Output sanitisation for tool responses.
 *
 *   - Sensitive keys (api_key, mnemonic, webhook_secret, …) have their
 *     values replaced with `"[REDACTED]"`.
 *   - String values longer than `MAX_STR` are truncated with a visible
 *     suffix — defends against prompt-injection payloads smuggled via
 *     attacker-controlled blockchain data (§4.4 of ALGOVOI_MCP.md).
 *
 * The public checkout `token` is NOT in the sensitive list: it is a
 * short opaque ID meant to be shown to the user, not a credential.
 */

export const MAX_STR = 512;

export const SENSITIVE_KEYS: ReadonlySet<string> = new Set([
  "mnemonic",
  "private_key",
  "privatekey",
  "secret",
  "api_key",
  "apikey",
  "password",
  "passwd",
  "authorization",
  "auth",
  "webhook_secret",
  "webhooksecret",
  "access_token",
  "refresh_token",
  "bearer_token",
  "algovoi_api_key",
  "algovoi_webhook_secret",
]);

const REDACTED = "[REDACTED]";

export function scrub<T>(obj: T): T {
  return _scrub(obj) as T;
}

function _scrub(obj: unknown): unknown {
  if (obj === null || obj === undefined) return obj;

  if (Array.isArray(obj)) {
    return obj.map(_scrub);
  }

  if (typeof obj === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj as Record<string, unknown>)) {
      if (SENSITIVE_KEYS.has(k.toLowerCase())) {
        out[k] = REDACTED;
      } else {
        out[k] = _scrub(v);
      }
    }
    return out;
  }

  if (typeof obj === "string" && obj.length > MAX_STR) {
    const extra = obj.length - MAX_STR;
    return obj.slice(0, MAX_STR) + `... [truncated ${extra} chars]`;
  }

  return obj;
}
